Change GitHub releases utility

see https://github.com/PhoenicisOrg/scripts/blob/master/Utils/Functions/Net/GithubReleases/script.js

The approach is wrong. Some scripts assumes that the download URL is formed in a certain way based on the release version. It would be better to return the complete json and then use `tarball_url` directly.

My suggestion would be to replace
```Javascript
const versions = JSON.parse(cat(releasesFile)).map(version => version.tag_name);
```
by
```Javascript
const versions = JSON.parse(cat(releasesFile));
```

I've tried it on local repo and I get this error:
```
[ERROR] org.phoenicis.multithreading.ControlledThreadPoolExecutorService (l.64) - Download of https://github.com/doitsujin/dxvk/releases/download/[object Object]/dxvk-[object Object].tar.gz has failed
	at org.phoenicis.tools.http.Downloader.saveConnectionToStream(Downloader.java:307)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:260)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:237)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:101)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:81)
	at <js> get(Unnamed:135-144:3538-3839)
	at <js> get(Unnamed:110-117:2786-3026)
	at <js> go(Unnamed:70-76:2648-2915)
	at <js> :anonymous(Unnamed:20:661-679)
	at <js> go(Unnamed:83:3245-3279)
	at com.oracle.truffle.polyglot.ObjectProxyHandler.invoke(HostInteropReflect.java:678)
	at com.sun.proxy.$Proxy68.go(Unknown Source)
	at org.phoenicis.javafx.components.application.skin.ApplicationInformationPanelSkin.lambda$installScript$7(ApplicationInformationPanelSkin.java:237)
	at org.phoenicis.scripts.session.PhoenicisInteractiveScriptSession.eval(PhoenicisInteractiveScriptSession.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$createInteractiveSession$1(BackgroundScriptInterpreter.java:45)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by host exception: org.phoenicis.tools.http.DownloadException: Download of https://github.com/doitsujin/dxvk/releases/download/[object Object]/dxvk-[object Object].tar.gz has failed

[WARNING] 
org.graalvm.polyglot.PolyglotException: Download of https://github.com/doitsujin/dxvk/releases/download/[object Object]/dxvk-[object Object].tar.gz has failed
    at org.phoenicis.tools.http.Downloader.saveConnectionToStream (Downloader.java:307)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:260)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:237)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:101)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:81)
    at <js>.get (Unnamed:135)
    at <js>.get (Unnamed:110)
    at <js>.go (Unnamed:70)
    at <js>.:anonymous (Unnamed:20)
    at <js>.go (Unnamed:83)
    at com.oracle.truffle.polyglot.ObjectProxyHandler.invoke (HostInteropReflect.java:678)
    at com.sun.proxy.$Proxy68.go (Unknown Source)
    at org.phoenicis.javafx.components.application.skin.ApplicationInformationPanelSkin.lambda$installScript$7 (ApplicationInformationPanelSkin.java:237)
    at org.phoenicis.scripts.session.PhoenicisInteractiveScriptSession.eval (PhoenicisInteractiveScriptSession.java:35)
    at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$createInteractiveSession$1 (BackgroundScriptInterpreter.java:45)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
```
I guess we need the `map` object after all because the version number is stored there.

However, I just tried installing DXVK in `Space Engineers` prefix via verb and got this error:
```
[ERROR] org.phoenicis.multithreading.ControlledThreadPoolExecutorService (l.64) - Download of https://github.com/doitsujin/dxvk/releases/download/v1.5/dxvk-v1.5.tar.gz has failed
	at org.phoenicis.tools.http.Downloader.saveConnectionToStream(Downloader.java:307)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:260)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:237)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:101)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:81)
	at <js> get(Unnamed:135-144:3538-3839)
	at <js> get(Unnamed:110-117:2786-3026)
	at <js> go(Unnamed:70-76:2648-2915)
	at <js> install(Unnamed:119:4293-4345)
	at org.graalvm.polyglot.Value.invokeMember(Value.java:459)
	at org.phoenicis.engines.VerbsManager.lambda$installVerb$0(VerbsManager.java:71)
	at org.phoenicis.scripts.session.PhoenicisInteractiveScriptSession.eval(PhoenicisInteractiveScriptSession.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$createInteractiveSession$1(BackgroundScriptInterpreter.java:45)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by host exception: org.phoenicis.tools.http.DownloadException: Download of https://github.com/doitsujin/dxvk/releases/download/v1.5/dxvk-v1.5.tar.gz has failed

[WARNING] 
org.graalvm.polyglot.PolyglotException: Download of https://github.com/doitsujin/dxvk/releases/download/v1.5/dxvk-v1.5.tar.gz has failed
    at org.phoenicis.tools.http.Downloader.saveConnectionToStream (Downloader.java:307)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:260)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:237)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:101)
    at org.phoenicis.tools.http.Downloader.get (Downloader.java:81)
    at <js>.get (Unnamed:135)
    at <js>.get (Unnamed:110)
    at <js>.go (Unnamed:70)
    at <js>.install (Unnamed:119)
    at org.graalvm.polyglot.Value.invokeMember (Value.java:459)
    at org.phoenicis.engines.VerbsManager.lambda$installVerb$0 (VerbsManager.java:71)
    at org.phoenicis.scripts.session.PhoenicisInteractiveScriptSession.eval (PhoenicisInteractiveScriptSession.java:35)
    at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$createInteractiveSession$1 (BackgroundScriptInterpreter.java:45)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
```
Tried D9VK on `The Sims 3` prefix and no error whatsoever.

After applying #1152 and updating POL5 to the latest commit the error when using DXVK verb have changed:
```
[ERROR] org.phoenicis.multithreading.ControlledThreadPoolExecutorService (l.64) - TypeError: invokeMember (menu) on JavaObject[org.phoenicis.scripts.wizard.UiSetupWizardImplementation@707ce5da (org.phoenicis.scripts.wizard.UiSetupWizardImplementation)] failed due to: no applicable overload found (overloads: [Method[public org.phoenicis.scripts.ui.MenuItem org.phoenicis.scripts.wizard.UiSetupWizardImplementation.menu(java.lang.String,java.util.List,java.lang.String)], Method[public org.phoenicis.scripts.ui.MenuItem org.phoenicis.scripts.wizard.UiSetupWizardImplementation.menu(java.lang.String,java.util.List)]], arguments: [Please select the version. (String), DynamicObject<JSArray>@35ce3283 (DynamicObjectBasic), DynamicObject<JSUserObject>@f36587d (DynamicObjectBasic)])
	at <js> install(Unnamed:116:4178-4245)
	at org.graalvm.polyglot.Value.invokeMember(Value.java:459)
	at org.phoenicis.engines.VerbsManager.lambda$installVerb$0(VerbsManager.java:71)
	at org.phoenicis.scripts.session.PhoenicisInteractiveScriptSession.eval(PhoenicisInteractiveScriptSession.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$createInteractiveSession$1(BackgroundScriptInterpreter.java:45)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
The same error is displayed when trying D9VK

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -5,7 +5,7 @@ const { ls, cp, remove } = include("utils.functions.filesystem.files");
 const operatingSystemFetcher = Bean("operatingSystemFetcher");
 const Optional = Java.type("java.util.Optional");
 const OverrideDLL = include("engines.wine.plugins.override_dll");
-const { getGithubReleases } = include("utils.functions.net.githubreleases");
+const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
 
 /**
  * Verb to install D9VK
@@ -28,6 +28,16 @@ class D9VK {
         return this;
     }
 
+    selectGithubVersion(wizard) {
+        const versions = fetchGithubReleases("Joshua-Ashton", "d9vk", wizard);
+
+        if (!this.d9vkVersion || typeof this.d9vkVersion !== 'string') {
+            return versions[0];
+        } else {
+            return versions.find(version => version.name === this.d9vkVersion);
+        }
+    }
+
     go() {
         const wizard = this.wine.wizard();
         const prefixDirectory = this.wine.prefixDirectory();
@@ -36,8 +46,7 @@ class D9VK {
 
         print("NOTE: Wine version should be greater or equal to 3.10");
 
-        if (operatingSystemFetcher.fetchCurrentOperationSystem().getFullName() !== "Linux")
-        {
+        if (operatingSystemFetcher.fetchCurrentOperationSystem().getFullName() !== "Linux") {
             const question = tr("D9VK is currently unsupported on non-Linux operating systems due to MoltenVK implementation being incomplete. Select how do you want to approach this situation.")
             const choices = [
                 tr("YES, continue with DXVK installation regardless"),
@@ -55,27 +64,16 @@ class D9VK {
                     // choice: "Exit D9VK Installer, but continue with the script"
                     return this;
                 default:
-                    // do nothing
+                // do nothing
             }
         }
-        else
-        {
+        else {
             wizard.message(tr("Please ensure you have the latest drivers (418.30 minimum for NVIDIA and mesa 19 for AMD) or else D9VK might not work correctly."));
         }
 
-        if (typeof this.d9vkVersion !== 'string')
-        {
-            const versions = getGithubReleases("Joshua-Ashton", "d9vk", wizard);
-            this.d9vkVersion = versions[0];
-        }
+        const selectedVersion = this.selectGithubVersion(wizard);
 
-        var setupFile = new Resource()
-            .wizard(wizard)
-            .url(
-                `https://github.com/Joshua-Ashton/d9vk/releases/download/${this.d9vkVersion}/d9vk-${this.d9vkVersion}.tar.gz`
-            )
-            .name(`d9vk-${this.d9vkVersion}.tar.gz`)
-            .get();
+        var setupFile = this.downloadGithubRelease(selectedVersion, wizard);
 
         new Extractor()
             .wizard(wizard)
@@ -83,7 +81,7 @@ class D9VK {
             .to(`${prefixDirectory}/TMP/`)
             .extract();
 
-        const d9vkTmpDir = `${prefixDirectory}/TMP/d9vk-${this.d9vkVersion}`;
+        const d9vkTmpDir = `${prefixDirectory}/TMP/d9vk-${selectedVersion.name}`;
 
         // copy 32 bits dll to system* and apply override
         ls(`${d9vkTmpDir}/x32`).forEach(file => {
@@ -115,8 +113,10 @@ class D9VK {
         wine.wizard(wizard);
         wine.prefix(container);
 
-        const versions = getGithubReleases("Joshua-Ashton", "d9vk", wizard);
-        const selectedVersion = wizard.menu(tr("Please select the version."), versions, versions[0]);
+        const versions = fetchGithubReleases("Joshua-Ashton", "d9vk", wizard);
+        const versionStrings = extractGithubReleaseStrings(versions);
+
+        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
 
         // install selected version
         new D9VK(wine).withVersion(selectedVersion.text).go();

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -1,5 +1,4 @@
 const Wine = include("engines.wine.engine.object");
-const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const { ls, cp, remove } = include("utils.functions.filesystem.files");
 const operatingSystemFetcher = Bean("operatingSystemFetcher");

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -5,7 +5,7 @@ const { ls, cp, remove } = include("utils.functions.filesystem.files");
 const operatingSystemFetcher = Bean("operatingSystemFetcher");
 const Optional = Java.type("java.util.Optional");
 const OverrideDLL = include("engines.wine.plugins.override_dll");
-const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
+const { GitHubReleaseDownloader } = include("utils.functions.net.githubreleases");
 
 /**
  * Verb to install D9VK
@@ -26,16 +26,6 @@ class D9VK {
         this.d9vkVersion = d9vkVersion;
 
         return this;
-    }
-
-    selectGithubVersion(wizard) {
-        const versions = fetchGithubReleases("Joshua-Ashton", "d9vk", wizard);
-
-        if (!this.d9vkVersion || typeof this.d9vkVersion !== 'string') {
-            return versions[0];
-        } else {
-            return versions.find(version => version.name === this.d9vkVersion);
-        }
     }
 
     go() {
@@ -71,9 +61,15 @@ class D9VK {
             wizard.message(tr("Please ensure you have the latest drivers (418.30 minimum for NVIDIA and mesa 19 for AMD) or else D9VK might not work correctly."));
         }
 
-        const selectedVersion = this.selectGithubVersion(wizard);
+        const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        var setupFile = downloadGithubRelease(selectedVersion, wizard);
+        if (typeof this.d9vkVersion !== 'string') {
+            this.d9vkVersion = githubDownloader.getLatestRelease();
+        }
+
+        const setupFile = githubDownloader.download(this.d9vkVersion);
 
         new Extractor()
             .wizard(wizard)
@@ -81,7 +77,8 @@ class D9VK {
             .to(`${prefixDirectory}/TMP/`)
             .extract();
 
-        const d9vkTmpDir = `${prefixDirectory}/TMP/d9vk-${selectedVersion.name}`;
+
+        const d9vkTmpDir = `${prefixDirectory}/TMP/d9vk-${this.d9vkVersion}`;
 
         // copy 32 bits dll to system* and apply override
         ls(`${d9vkTmpDir}/x32`).forEach(file => {
@@ -113,10 +110,14 @@ class D9VK {
         wine.wizard(wizard);
         wine.prefix(container);
 
-        const versions = fetchGithubReleases("Joshua-Ashton", "d9vk", wizard);
-        const versionStrings = extractGithubReleaseStrings(versions);
+        const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
+        const versions = githubDownloader.getReleases();
+        const latestVersion = githubDownloader.getLatestRelease();
+
+        const selectedVersion = wizard.menu(tr("Please select the version."), versions, latestVersion);
 
         // install selected version
         new D9VK(wine).withVersion(selectedVersion.text).go();

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -73,7 +73,7 @@ class D9VK {
 
         const selectedVersion = this.selectGithubVersion(wizard);
 
-        var setupFile = this.downloadGithubRelease(selectedVersion, wizard);
+        var setupFile = downloadGithubRelease(selectedVersion, wizard);
 
         new Extractor()
             .wizard(wizard)

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -68,7 +68,7 @@ class D9VK {
             this.d9vkVersion = githubDownloader.getLatestRelease();
         }
 
-        const setupFile = githubDownloader.download(this.d9vkVersion);
+        const [setupFile] = githubDownloader.download(this.d9vkVersion);
 
         new Extractor()
             .wizard(wizard)

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -61,8 +61,9 @@ class D9VK {
         }
 
         const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         if (typeof this.d9vkVersion !== 'string') {
             this.d9vkVersion = githubDownloader.getLatestRelease();
@@ -110,8 +111,9 @@ class D9VK {
         wine.prefix(container);
 
         const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/D9VK/script.js
+++ b/Engines/Wine/Verbs/D9VK/script.js
@@ -60,10 +60,7 @@ class D9VK {
             wizard.message(tr("Please ensure you have the latest drivers (418.30 minimum for NVIDIA and mesa 19 for AMD) or else D9VK might not work correctly."));
         }
 
-        const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk", wizard);
 
         if (typeof this.d9vkVersion !== 'string') {
             this.d9vkVersion = githubDownloader.getLatestRelease();
@@ -110,10 +107,7 @@ class D9VK {
         wine.wizard(wizard);
         wine.prefix(container);
 
-        const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("Joshua-Ashton", "d9vk", wizard);
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/FAudio/script.js
+++ b/Engines/Wine/Verbs/FAudio/script.js
@@ -1,5 +1,4 @@
 const Wine = include("engines.wine.engine.object");
-const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const { ls, cp } = include("utils.functions.filesystem.files");
 const { GitHubReleaseDownloader } = include("utils.functions.net.githubreleases");

--- a/Engines/Wine/Verbs/FAudio/script.js
+++ b/Engines/Wine/Verbs/FAudio/script.js
@@ -45,7 +45,7 @@ class FAudio {
             this.faudioVersion = githubDownloader.getLatestRelease();
         }
 
-        const setupFile = githubDownloader.download(this.faudioVersion);
+        const [setupFile] = githubDownloader.download(this.faudioVersion);
 
         new Extractor()
             .wizard(wizard)

--- a/Engines/Wine/Verbs/FAudio/script.js
+++ b/Engines/Wine/Verbs/FAudio/script.js
@@ -37,10 +37,7 @@ class FAudio {
             throw "FAudio does not support 32bit architecture.";
         }
 
-        const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds", wizard);
 
         if (typeof this.faudioVersion !== "string") {
             this.faudioVersion = githubDownloader.getLatestRelease();
@@ -72,10 +69,7 @@ class FAudio {
         wine.prefix(container);
         wine.wizard(wizard);
 
-        const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds", wizard);
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/FAudio/script.js
+++ b/Engines/Wine/Verbs/FAudio/script.js
@@ -2,7 +2,7 @@ const Wine = include("engines.wine.engine.object");
 const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const { ls, cp } = include("utils.functions.filesystem.files");
-const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
+const { GitHubReleaseDownloader } = include("utils.functions.net.githubreleases");
 
 const Optional = Java.type("java.util.Optional");
 
@@ -29,16 +29,6 @@ class FAudio {
         return this;
     }
 
-    selectGithubVersion(wizard) {
-        const versions = fetchGithubReleases("Kron4ek", "FAudio-Builds", wizard);
-
-        if (!this.faudioVersion || typeof this.faudioVersion !== 'string') {
-            return versions[0];
-        } else {
-            return versions.find(version => version.name === this.faudioVersion);
-        }
-    }
-
     go() {
         const wizard = this.wine.wizard();
         const prefixDirectory = this.wine.prefixDirectory();
@@ -48,9 +38,15 @@ class FAudio {
             throw "FAudio does not support 32bit architecture.";
         }
 
-        const selectedVersion = this.selectGithubVersion(wizard);
+        const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        const setupFile = downloadGithubRelease(selectedVersion, wizard);
+        if (typeof this.faudioVersion !== "string") {
+            this.faudioVersion = githubDownloader.getLatestRelease();
+        }
+
+        const setupFile = githubDownloader.download(this.faudioVersion);
 
         new Extractor()
             .wizard(wizard)
@@ -58,7 +54,7 @@ class FAudio {
             .to(`${prefixDirectory}/FAudio/`)
             .extract();
 
-        const faudioDir = `${prefixDirectory}/FAudio/faudio-${selectedVersion.name}`;
+        const faudioDir = `${prefixDirectory}/FAudio/faudio-${this.faudioVersion}`;
 
         ls(`${faudioDir}/x64`).forEach(file => {
             if (file.endsWith(".dll")) {
@@ -76,10 +72,14 @@ class FAudio {
         wine.prefix(container);
         wine.wizard(wizard);
 
-        const versions = fetchGithubReleases("Kron4ek", "FAudio-Builds", wizard);
-        const versionStrings = extractGithubReleaseStrings(versions);
+        const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
+        const versions = githubDownloader.getReleases();
+        const latestVersion = githubDownloader.getLatestRelease();
+
+        const selectedVersion = wizard.menu(tr("Please select the version."), versions, latestVersion);
 
         // install selected version
         new FAudio(wine).withVersion(selectedVersion.text).go();

--- a/Engines/Wine/Verbs/FAudio/script.js
+++ b/Engines/Wine/Verbs/FAudio/script.js
@@ -38,8 +38,9 @@ class FAudio {
         }
 
         const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         if (typeof this.faudioVersion !== "string") {
             this.faudioVersion = githubDownloader.getLatestRelease();
@@ -72,8 +73,9 @@ class FAudio {
         wine.wizard(wizard);
 
         const githubDownloader = new GitHubReleaseDownloader("Kron4ek", "FAudio-Builds")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/FAudio/script.js
+++ b/Engines/Wine/Verbs/FAudio/script.js
@@ -2,7 +2,7 @@ const Wine = include("engines.wine.engine.object");
 const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const { ls, cp } = include("utils.functions.filesystem.files");
-const { getGithubReleases } = include("utils.functions.net.githubreleases");
+const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
 
 const Optional = Java.type("java.util.Optional");
 
@@ -29,6 +29,16 @@ class FAudio {
         return this;
     }
 
+    selectGithubVersion(wizard) {
+        const versions = fetchGithubReleases("Kron4ek", "FAudio-Builds", wizard);
+
+        if (!this.faudioVersion || typeof this.faudioVersion !== 'string') {
+            return versions[0];
+        } else {
+            return versions.find(version => version.name === this.faudioVersion);
+        }
+    }
+
     go() {
         const wizard = this.wine.wizard();
         const prefixDirectory = this.wine.prefixDirectory();
@@ -38,18 +48,9 @@ class FAudio {
             throw "FAudio does not support 32bit architecture.";
         }
 
-        if (typeof this.faudioVersion !== "string") {
-            const versions = getGithubReleases("Kron4ek", "FAudio-Builds", wizard);
-            this.faudioVersion = versions[0];
-        }
+        const selectedVersion = this.selectGithubVersion(wizard);
 
-        const setupFile = new Resource()
-            .wizard(wizard)
-            .url(
-                `https://github.com/Kron4ek/FAudio-Builds/releases/download/${this.faudioVersion}/faudio-${this.faudioVersion}.tar.xz`
-            )
-            .name(`faudio-${this.faudioVersion}.tar.xz`)
-            .get();
+        const setupFile = downloadGithubRelease(selectedVersion, wizard);
 
         new Extractor()
             .wizard(wizard)
@@ -57,7 +58,7 @@ class FAudio {
             .to(`${prefixDirectory}/FAudio/`)
             .extract();
 
-        const faudioDir = `${prefixDirectory}/FAudio/faudio-${this.faudioVersion}`;
+        const faudioDir = `${prefixDirectory}/FAudio/faudio-${selectedVersion.name}`;
 
         ls(`${faudioDir}/x64`).forEach(file => {
             if (file.endsWith(".dll")) {
@@ -75,9 +76,10 @@ class FAudio {
         wine.prefix(container);
         wine.wizard(wizard);
 
-        const versions = getGithubReleases("Kron4ek", "FAudio-Builds", wizard);
+        const versions = fetchGithubReleases("Kron4ek", "FAudio-Builds", wizard);
+        const versionStrings = extractGithubReleaseStrings(versions);
 
-        const selectedVersion = wizard.menu(tr("Please select the version."), versions, versions[0]);
+        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
 
         // install selected version
         new FAudio(wine).withVersion(selectedVersion.text).go();

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -4,6 +4,7 @@ const operatingSystemFetcher = Bean("operatingSystemFetcher");
 const Optional = Java.type("java.util.Optional");
 const OverrideDLL = include("engines.wine.plugins.override_dll");
 const { GitHubReleaseDownloader } = include("utils.functions.net.githubreleases");
+const { cp, remove } = include("utils.functions.filesystem.files");
 
 /**
  * Verb to install VK9

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -1,5 +1,4 @@
 const Wine = include("engines.wine.engine.object");
-const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const operatingSystemFetcher = Bean("operatingSystemFetcher");
 const Optional = Java.type("java.util.Optional");

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -4,7 +4,7 @@ const { Extractor } = include("utils.functions.filesystem.extract");
 const operatingSystemFetcher = Bean("operatingSystemFetcher");
 const Optional = Java.type("java.util.Optional");
 const OverrideDLL = include("engines.wine.plugins.override_dll");
-const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
+const { GitHubReleaseDownloader } = include("utils.functions.net.githubreleases");
 
 /**
  * Verb to install VK9
@@ -25,16 +25,6 @@ class VK9 {
         this.vk9Version = vk9Version;
 
         return this;
-    }
-
-    selectGithubVersion(wizard) {
-        const versions = fetchGithubReleases("disks86", "VK9", wizard);
-
-        if (!this.vk9Version || typeof this.vk9Version !== 'string') {
-            return versions[0];
-        } else {
-            return versions.find(version => version.name === this.vk9Version);
-        }
     }
 
     go() {
@@ -68,9 +58,15 @@ class VK9 {
         print("NOTE: wine version should be greater or equal to 3.5");
         print("NOTE: works from 0.28.0");
 
-        const selectedVersion = this.selectGithubVersion(wizard);
+        const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        const setupFile32 = downloadGithubRelease(selectedVersion, wizard);
+        if (typeof this.vk9Version !== "string") {
+            this.vk9Version = githubDownloader.getLatestRelease();
+        }
+
+        const setupFile32 = githubDownloader.download(this.vk9Version, /(.+)-bin-x86-Release.zip/);
 
         new Extractor()
             .wizard(wizard)
@@ -78,18 +74,12 @@ class VK9 {
             .to(`${prefixDirectory}/TMP32/`)
             .extract();
 
-        cp(`${prefixDirectory}/TMP32/${selectedVersion.name}-bin-x86-Release/d3d9.dll`, system32directory);
+        cp(`${prefixDirectory}/TMP32/${this.vk9Version}-bin-x86-Release/d3d9.dll`, system32directory);
 
         remove(`${prefixDirectory}/TMP32/`);
 
         if (this.wine.architecture() === "amd64") {
-            const setupFile64 = new Resource()
-                .wizard(wizard)
-                .url(
-                    `https://github.com/disks86/VK9/releases/download/${selectedVersion.name}/${selectedVersion.name}-bin-x86_64-Release.zip`
-                )
-                .name(`${selectedVersion.name}-bin-x86_64-Realease.zip`)
-                .get();
+            const setupFile64 = githubDownloader.download(this.vk9Version, /(.+)-bin-x86_64-Release.zip/);
 
             new Extractor()
                 .wizard(wizard)
@@ -97,7 +87,7 @@ class VK9 {
                 .to(`${prefixDirectory}/TMP64/`)
                 .extract();
 
-            cp(`${prefixDirectory}/TMP64/${selectedVersion.name}-bin-x86_64-Release/d3d9.dll`, system64directory);
+            cp(`${prefixDirectory}/TMP64/${this.vk9Version}-bin-x86_64-Release/d3d9.dll`, system64directory);
 
             remove(`${prefixDirectory}/TMP64/`);
         }
@@ -112,10 +102,14 @@ class VK9 {
         wine.wizard(wizard);
         wine.prefix(container);
 
-        const versions = fetchGithubReleases("disks86", "VK9", wizard);
-        const versionStrings = extractGithubReleaseStrings(versions);
+        const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
+        const versions = githubDownloader.getReleases();
+        const latestVersion = githubDownloader.getLatestRelease();
+
+        const selectedVersion = wizard.menu(tr("Please select the version."), versions, latestVersion);
 
         // install selected version
         new VK9(wine).withVersion(selectedVersion.text).go();

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -65,7 +65,7 @@ class VK9 {
             this.vk9Version = githubDownloader.getLatestRelease();
         }
 
-        const setupFile32 = githubDownloader.download(this.vk9Version, /(.+)-bin-x86-Release.zip/);
+        const [setupFile32] = githubDownloader.download(this.vk9Version, /(.+)-bin-x86-Release.zip/);
 
         new Extractor()
             .wizard(wizard)
@@ -78,7 +78,7 @@ class VK9 {
         remove(`${prefixDirectory}/TMP32/`);
 
         if (this.wine.architecture() === "amd64") {
-            const setupFile64 = githubDownloader.download(this.vk9Version, /(.+)-bin-x86_64-Release.zip/);
+            const [setupFile64] = githubDownloader.download(this.vk9Version, /(.+)-bin-x86_64-Release.zip/);
 
             new Extractor()
                 .wizard(wizard)

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -4,7 +4,7 @@ const { Extractor } = include("utils.functions.filesystem.extract");
 const operatingSystemFetcher = Bean("operatingSystemFetcher");
 const Optional = Java.type("java.util.Optional");
 const OverrideDLL = include("engines.wine.plugins.override_dll");
-const { getGithubReleases } = include("utils.functions.net.githubreleases");
+const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
 
 /**
  * Verb to install VK9
@@ -27,14 +27,23 @@ class VK9 {
         return this;
     }
 
+    selectGithubVersion(wizard) {
+        const versions = fetchGithubReleases("disks86", "VK9", wizard);
+
+        if (!this.vk9Version || typeof this.vk9Version !== 'string') {
+            return versions[0];
+        } else {
+            return versions.find(version => version.name === this.vk9Version);
+        }
+    }
+
     go() {
         const wizard = this.wine.wizard();
         const prefixDirectory = this.wine.prefixDirectory();
         const system32directory = this.wine.system32directory();
         const system64directory = this.wine.system64directory();
 
-        if (operatingSystemFetcher.fetchCurrentOperationSystem().getFullName() !== "Linux")
-        {
+        if (operatingSystemFetcher.fetchCurrentOperationSystem().getFullName() !== "Linux") {
             const question = tr("VK9 is currently unsupported on non-Linux operating systems due to MoltenVK implementation being incomplete. Select how do you want to approach this situation.")
             const choices = [
                 tr("YES, continue with VK9 installation regardless"),
@@ -52,25 +61,16 @@ class VK9 {
                     // choice: "Exit VK9 Installer, but continue with the script"
                     return this;
                 default:
-                    // do nothing
+                // do nothing
             }
         }
 
         print("NOTE: wine version should be greater or equal to 3.5");
         print("NOTE: works from 0.28.0");
 
-        if (typeof this.vk9Version !== "string") {
-            const versions = getGithubReleases("disks86", "VK9", wizard);
-            this.vk9Version = versions[0];
-        }
+        const selectedVersion = this.selectGithubVersion(wizard);
 
-        const setupFile32 = new Resource()
-            .wizard(wizard)
-            .url(
-                `https://github.com/disks86/VK9/releases/download/${this.vk9Version}/${this.vk9Version}-bin-x86-Release.zip`
-            )
-            .name(`${this.vk9Version}-bin-x86-Realease.zip`)
-            .get();
+        const setupFile32 = downloadGithubRelease(selectedVersion, wizard);
 
         new Extractor()
             .wizard(wizard)
@@ -78,7 +78,7 @@ class VK9 {
             .to(`${prefixDirectory}/TMP32/`)
             .extract();
 
-        cp(`${prefixDirectory}/TMP32/${this.vk9Version}-bin-x86-Release/d3d9.dll`, system32directory);
+        cp(`${prefixDirectory}/TMP32/${selectedVersion.name}-bin-x86-Release/d3d9.dll`, system32directory);
 
         remove(`${prefixDirectory}/TMP32/`);
 
@@ -86,9 +86,9 @@ class VK9 {
             const setupFile64 = new Resource()
                 .wizard(wizard)
                 .url(
-                    `https://github.com/disks86/VK9/releases/download/${this.vk9Version}/${this.vk9Version}-bin-x86_64-Release.zip`
+                    `https://github.com/disks86/VK9/releases/download/${selectedVersion.name}/${selectedVersion.name}-bin-x86_64-Release.zip`
                 )
-                .name(`${this.vk9Version}-bin-x86_64-Realease.zip`)
+                .name(`${selectedVersion.name}-bin-x86_64-Realease.zip`)
                 .get();
 
             new Extractor()
@@ -97,7 +97,7 @@ class VK9 {
                 .to(`${prefixDirectory}/TMP64/`)
                 .extract();
 
-            cp(`${prefixDirectory}/TMP64/${this.vk9Version}-bin-x86_64-Release/d3d9.dll`, system64directory);
+            cp(`${prefixDirectory}/TMP64/${selectedVersion.name}-bin-x86_64-Release/d3d9.dll`, system64directory);
 
             remove(`${prefixDirectory}/TMP64/`);
         }
@@ -112,8 +112,10 @@ class VK9 {
         wine.wizard(wizard);
         wine.prefix(container);
 
-        const versions = getGithubReleases("disks86", "VK9", wizard);
-        const selectedVersion = wizard.menu(tr("Please select the version."), versions, versions[0]);
+        const versions = fetchGithubReleases("disks86", "VK9", wizard);
+        const versionStrings = extractGithubReleaseStrings(versions);
+
+        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
 
         // install selected version
         new VK9(wine).withVersion(selectedVersion.text).go();

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -59,10 +59,7 @@ class VK9 {
         print("NOTE: wine version should be greater or equal to 3.5");
         print("NOTE: works from 0.28.0");
 
-        const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9", wizard);
 
         if (typeof this.vk9Version !== "string") {
             this.vk9Version = githubDownloader.getLatestRelease();
@@ -104,10 +101,7 @@ class VK9 {
         wine.wizard(wizard);
         wine.prefix(container);
 
-        const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9", wizard);
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/VK9/script.js
+++ b/Engines/Wine/Verbs/VK9/script.js
@@ -58,8 +58,9 @@ class VK9 {
         print("NOTE: works from 0.28.0");
 
         const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         if (typeof this.vk9Version !== "string") {
             this.vk9Version = githubDownloader.getLatestRelease();
@@ -102,8 +103,9 @@ class VK9 {
         wine.prefix(container);
 
         const githubDownloader = new GitHubReleaseDownloader("disks86", "VK9")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/gallium9/script.js
+++ b/Engines/Wine/Verbs/gallium9/script.js
@@ -29,22 +29,10 @@ class Gallium9 {
         return this;
     }
 
-    selectGithubVersion(wizard) {
-        const versions = fetchGithubReleases("iXit", "wine-nine-standalone", wizard);
-
-        if (!this.gallium9Version || typeof this.gallium9Version !== 'string') {
-            return versions[0];
-        } else {
-            return versions.find(version => version.name === this.gallium9Version);
-        }
-    }
-
     go() {
         const wizard = this.wine.wizard();
         const prefixDirectory = this.wine.prefixDirectory();
         const system32directory = this.wine.system32directory();
-
-        const selectedVersion = this.selectGithubVersion(wizard);
 
         wizard.message(
             tr(
@@ -52,7 +40,15 @@ class Gallium9 {
             )
         );
 
-        const setupFile = downloadGithubRelease(selectedVersion, wizard);
+        const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone")
+            .withWizard(wizard)
+            .fetchReleases();
+
+        if (typeof this.gallium9Version !== "string") {
+            this.gallium9Version = githubDownloader.getLatestRelease();
+        }
+
+        const setupFile = githubDownloader.download(this.gallium9Version);
 
         new Extractor()
             .wizard(wizard)
@@ -96,10 +92,14 @@ class Gallium9 {
         const wine = new Wine();
         const wizard = SetupWizard(InstallationType.VERBS, "gallium9", Optional.empty());
 
-        const versions = fetchGithubReleases("iXit", "wine-nine-standalone", wizard);
-        const versionStrings = extractGithubReleaseStrings(versions);
+        const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone")
+            .withWizard(wizard)
+            .fetchReleases();
 
-        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
+        const versions = githubDownloader.getReleases();
+        const latestVersion = githubDownloader.getLatestRelease();
+
+        const selectedVersion = wizard.menu(tr("Please select the version."), versions, latestVersion);
 
         wine.prefix(container);
         wine.wizard(wizard);

--- a/Engines/Wine/Verbs/gallium9/script.js
+++ b/Engines/Wine/Verbs/gallium9/script.js
@@ -40,8 +40,9 @@ class Gallium9 {
         );
 
         const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         if (typeof this.gallium9Version !== "string") {
             this.gallium9Version = githubDownloader.getLatestRelease();
@@ -92,8 +93,9 @@ class Gallium9 {
         const wizard = SetupWizard(InstallationType.VERBS, "gallium9", Optional.empty());
 
         const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone")
-            .withWizard(wizard)
-            .fetchReleases();
+            .withWizard(wizard);
+
+        githubDownloader.fetchReleases();
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/gallium9/script.js
+++ b/Engines/Wine/Verbs/gallium9/script.js
@@ -47,7 +47,7 @@ class Gallium9 {
             this.gallium9Version = githubDownloader.getLatestRelease();
         }
 
-        const setupFile = githubDownloader.download(this.gallium9Version);
+        const [setupFile] = githubDownloader.download(this.gallium9Version);
 
         new Extractor()
             .wizard(wizard)

--- a/Engines/Wine/Verbs/gallium9/script.js
+++ b/Engines/Wine/Verbs/gallium9/script.js
@@ -2,7 +2,7 @@ const Wine = include("engines.wine.engine.object");
 const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const { remove, lns } = include("utils.functions.filesystem.files");
-const { getGithubReleases } = include("utils.functions.net.githubreleases");
+const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
 
 const Optional = Java.type("java.util.Optional");
 
@@ -29,15 +29,22 @@ class Gallium9 {
         return this;
     }
 
+    selectGithubVersion(wizard) {
+        const versions = fetchGithubReleases("iXit", "wine-nine-standalone", wizard);
+
+        if (!this.gallium9Version || typeof this.gallium9Version !== 'string') {
+            return versions[0];
+        } else {
+            return versions.find(version => version.name === this.gallium9Version);
+        }
+    }
+
     go() {
         const wizard = this.wine.wizard();
         const prefixDirectory = this.wine.prefixDirectory();
         const system32directory = this.wine.system32directory();
 
-        if (typeof this.gallium9Version !== "string") {
-            const versions = getGithubReleases("iXit", "wine-nine-standalone", wizard);
-            this.gallium9Version = versions[0];
-        }
+        const selectedVersion = this.selectGithubVersion(wizard);
 
         wizard.message(
             tr(
@@ -45,13 +52,7 @@ class Gallium9 {
             )
         );
 
-        const setupFile = new Resource()
-            .wizard(wizard)
-            .url(
-                `https://github.com/iXit/wine-nine-standalone/releases/download/${this.gallium9Version}/gallium-nine-standalone-${this.gallium9Version}.tar.gz`
-            )
-            .name(`gallium-nine-standalone-${this.gallium9Version}.tar.gz`)
-            .get();
+        const setupFile = downloadGithubRelease(selectedVersion, wizard);
 
         new Extractor()
             .wizard(wizard)
@@ -95,9 +96,10 @@ class Gallium9 {
         const wine = new Wine();
         const wizard = SetupWizard(InstallationType.VERBS, "gallium9", Optional.empty());
 
-        const versions = getGithubReleases("iXit", "wine-nine-standalone", wizard);
+        const versions = fetchGithubReleases("iXit", "wine-nine-standalone", wizard);
+        const versionStrings = extractGithubReleaseStrings(versions);
 
-        const selectedVersion = wizard.menu(tr("Please select the version."), versions, versions[0]);
+        const selectedVersion = wizard.menu(tr("Please select the version."), versionStrings, versionStrings[0]);
 
         wine.prefix(container);
         wine.wizard(wizard);

--- a/Engines/Wine/Verbs/gallium9/script.js
+++ b/Engines/Wine/Verbs/gallium9/script.js
@@ -39,10 +39,7 @@ class Gallium9 {
             )
         );
 
-        const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone", wizard);
 
         if (typeof this.gallium9Version !== "string") {
             this.gallium9Version = githubDownloader.getLatestRelease();
@@ -92,10 +89,7 @@ class Gallium9 {
         const wine = new Wine();
         const wizard = SetupWizard(InstallationType.VERBS, "gallium9", Optional.empty());
 
-        const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone")
-            .withWizard(wizard);
-
-        githubDownloader.fetchReleases();
+        const githubDownloader = new GitHubReleaseDownloader("iXit", "wine-nine-standalone", wizard);
 
         const versions = githubDownloader.getReleases();
         const latestVersion = githubDownloader.getLatestRelease();

--- a/Engines/Wine/Verbs/gallium9/script.js
+++ b/Engines/Wine/Verbs/gallium9/script.js
@@ -1,8 +1,7 @@
 const Wine = include("engines.wine.engine.object");
-const Resource = include("utils.functions.net.resource");
 const { Extractor } = include("utils.functions.filesystem.extract");
 const { remove, lns } = include("utils.functions.filesystem.files");
-const { fetchGithubReleases, downloadGithubRelease, extractGithubReleaseStrings } = include("utils.functions.net.githubreleases");
+const { GitHubReleaseDownloader } = include("utils.functions.net.githubreleases");
 
 const Optional = Java.type("java.util.Optional");
 

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -7,8 +7,8 @@ const { createTempDir, remove, cat } = include("utils.functions.filesystem.files
  */
 module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
     /**
-     * Constructor 
-     * 
+     * Constructor
+     *
      * @param {string} repositoryOwner The owner of the GitHub repository
      * @param {string} repositoryName The name of the GitHub repository
      */
@@ -19,7 +19,7 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
 
     /**
      * Sets the setup wizard
-     * 
+     *
      * @param {SetupWizard} wizard The setup wizard
      * @returns {GitHubReleaseDownloader} this
      */
@@ -31,6 +31,8 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
 
     /**
      * Fetches the GitHub releases
+     * 
+     * @returns {void}
      */
     fetchReleases() {
         const tmpDir = createTempDir();
@@ -49,7 +51,7 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
 
     /**
      * Fetches the tags belonging to the releases for the given GitHub repository
-     * 
+     *
      * @returns {string[]} An array containing the tags for the GitHub releases
      */
     getReleases() {
@@ -58,7 +60,7 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
 
     /**
      * Fetches the tag belonging to the latest release for the given GitHub repository
-     * 
+     *
      * @returns {string} The tag belonging to the latest release
      */
     getLatestRelease() {
@@ -67,7 +69,7 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
 
     /**
      * Downloads all assets that belong to the release with the given tag that match a given regex
-     * 
+     *
      * @param {string} tag The tag name of the release
      * @param {RegExp} assetRegex An optional regex describing the desired asset(s)
      * @returns {string[]} An array of paths leading to the downloaded files

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -31,7 +31,7 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
 
     /**
      * Fetches the GitHub releases
-     * 
+     *
      * @returns {void}
      */
     fetchReleases() {

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -1,29 +1,41 @@
 const Downloader = include("utils.functions.net.download");
-const {createTempDir, remove, cat} = include("utils.functions.filesystem.files");
+const Resource = include("utils.functions.net.resource");
+const { createTempDir, remove, cat } = include("utils.functions.filesystem.files");
 
 /**
- * Get an array of tag corresponding to the releases in a git repository
+ * Fetches an array of tag corresponding to the releases in a git repository
  *
  * @param {string} repositoryOwner owner of the repository
  * @param {string} repositoryName name of the repository
  * @param {wizard} wizard the wizard
- * @returns {string[]} list of version tag in the repository
+ * @returns {object[]} list of version objects in the repository
  */
-module.getGithubReleases = function getGithubReleases(repositoryOwner, repositoryName, wizard) {
+module.fetchGithubReleases = function fetchGithubReleases(repositoryOwner, repositoryName, wizard) {
     const tmpDir = createTempDir();
 
     const releasesFile = new Downloader()
         .wizard(wizard)
-        .url(
-            `https://api.github.com/repos/${repositoryOwner}/${repositoryName}/releases`
-        )
+        .url(`https://api.github.com/repos/${repositoryOwner}/${repositoryName}/releases`)
         .message(tr("Fetching versions list..."))
         .to(tmpDir + "/releases.json")
         .get();
 
-    const versions = JSON.parse(cat(releasesFile)).map(version => version.tag_name);
+    const versions = JSON.parse(cat(releasesFile));
 
     remove(tmpDir);
 
     return versions;
 };
+
+module.extractGithubReleaseStrings = function extractGithubReleaseStrings(versions) {
+    return versions.map(version => version.name);
+}
+
+module.downloadGithubRelease = function downloadGithubRelease(version, wizard) {
+    return new Resource()
+        .wizard(wizard)
+        .url(version.tarball_url)
+        // TODO: generate a more "random" name to prevent file conflicts
+        .name(`${version.name}.tar.xz`)
+        .get();
+}

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -11,28 +11,20 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
      *
      * @param {string} repositoryOwner The owner of the GitHub repository
      * @param {string} repositoryName The name of the GitHub repository
+     * @param {SetupWizard} wizard The setup wizard
      */
-    constructor(repositoryOwner, repositoryName) {
+    constructor(repositoryOwner, repositoryName, wizard) {
         this.repositoryOwner = repositoryOwner;
         this.repositoryName = repositoryName;
-    }
-
-    /**
-     * Sets the setup wizard
-     *
-     * @param {SetupWizard} wizard The setup wizard
-     * @returns {GitHubReleaseDownloader} this
-     */
-    withWizard(wizard) {
         this.wizard = wizard;
 
-        return this;
+        this.versions = this.fetchReleases();
     }
 
     /**
      * Fetches the GitHub releases
      *
-     * @returns {void}
+     * @returns {object[]} The downloaded versions json array
      */
     fetchReleases() {
         const tmpDir = createTempDir();
@@ -44,9 +36,11 @@ module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
             .to(tmpDir + "/releases.json")
             .get();
 
-        this.versions = JSON.parse(cat(releasesFile));
+        const versions = JSON.parse(cat(releasesFile));
 
         remove(tmpDir);
+
+        return versions;
     }
 
     /**

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -2,40 +2,81 @@ const Downloader = include("utils.functions.net.download");
 const Resource = include("utils.functions.net.resource");
 const { createTempDir, remove, cat } = include("utils.functions.filesystem.files");
 
-/**
- * Fetches an array of tag corresponding to the releases in a git repository
- *
- * @param {string} repositoryOwner owner of the repository
- * @param {string} repositoryName name of the repository
- * @param {wizard} wizard the wizard
- * @returns {object[]} list of version objects in the repository
- */
-module.fetchGithubReleases = function fetchGithubReleases(repositoryOwner, repositoryName, wizard) {
-    const tmpDir = createTempDir();
+module.GitHubReleaseDownloader = class {
+    /**
+     * Constructor 
+     * 
+     * @param {string} repositoryOwner The owner of the GitHub repository
+     * @param {string} repositoryName The name of the GitHub repository
+     */
+    constructor(repositoryOwner, repositoryName) {
+        super();
 
-    const releasesFile = new Downloader()
-        .wizard(wizard)
-        .url(`https://api.github.com/repos/${repositoryOwner}/${repositoryName}/releases`)
-        .message(tr("Fetching versions list..."))
-        .to(tmpDir + "/releases.json")
-        .get();
+        this.repositoryOwner = repositoryOwner;
+        this.repositoryName = repositoryName;
+    }
 
-    const versions = JSON.parse(cat(releasesFile));
+    /**
+     * Sets the setup wizard
+     * 
+     * @param {SetupWizard} wizard The setup wizard
+     * @returns {GitHubReleaseDownloader} this
+     */
+    withWizard(wizard) {
+        this.wizard = wizard;
 
-    remove(tmpDir);
+        return this;
+    }
 
-    return versions;
+    /**
+     * Fetches the GitHub releases
+     */
+    fetchReleases() {
+        const tmpDir = createTempDir();
+
+        const releasesFile = new Downloader()
+            .wizard(this.wizard)
+            .url(`https://api.github.com/repos/${this.repositoryOwner}/${this.repositoryName}/releases`)
+            .message(tr("Fetching versions list..."))
+            .to(tmpDir + "/releases.json")
+            .get();
+
+        this.versions = JSON.parse(cat(releasesFile));
+
+        remove(tmpDir);
+    }
+
+    /**
+     * Fetches the tags belonging to the releases for the given GitHub repository
+     * 
+     * @returns {string[]} An array containing the tags for the GitHub releases
+     */
+    getReleases() {
+        return this.versions.map(version => version.tag_name);
+    }
+
+    getLatestRelease() {
+        return this.versions[0].tag_name;
+    }
+
+    /**
+     * Downloads all assets that belong to the release with the given tag that match a given regex
+     * 
+     * @param {string} tag The tag name of the release
+     * @param {RegExp} assetRegex An optional regex describing the desired asset(s)
+     * @returns {string[]} An array of paths leading to the downloaded files
+     */
+    download(tag, assetRegex) {
+        const version = this.versions.find(version => version.tag_name === tag);
+
+        return version.assets
+            .filter(asset => !assetRegex || asset.name.match(assetRegex))
+            .map(asset =>
+                new Resource()
+                    .wizard(this.wizard)
+                    .url(asset.url)
+                    .name(asset.name)
+                    .get()
+            );
+    }
 };
-
-module.extractGithubReleaseStrings = function extractGithubReleaseStrings(versions) {
-    return versions.map(version => version.name);
-}
-
-module.downloadGithubRelease = function downloadGithubRelease(version, wizard) {
-    return new Resource()
-        .wizard(wizard)
-        .url(version.tarball_url)
-        // TODO: generate a more "random" name to prevent file conflicts
-        .name(`${version.name}.tar.xz`)
-        .get();
-}

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -2,7 +2,7 @@ const Downloader = include("utils.functions.net.download");
 const Resource = include("utils.functions.net.resource");
 const { createTempDir, remove, cat } = include("utils.functions.filesystem.files");
 
-module.GitHubReleaseDownloader = class {
+module.GitHubReleaseDownloader = class GitHubReleaseDownloader {
     /**
      * Constructor 
      * 
@@ -10,8 +10,6 @@ module.GitHubReleaseDownloader = class {
      * @param {string} repositoryName The name of the GitHub repository
      */
     constructor(repositoryOwner, repositoryName) {
-        super();
-
         this.repositoryOwner = repositoryOwner;
         this.repositoryName = repositoryName;
     }

--- a/Utils/Functions/Net/GithubReleases/script.js
+++ b/Utils/Functions/Net/GithubReleases/script.js
@@ -74,7 +74,7 @@ module.GitHubReleaseDownloader = class {
             .map(asset =>
                 new Resource()
                     .wizard(this.wizard)
-                    .url(asset.url)
+                    .url(asset.browser_download_url)
                     .name(asset.name)
                     .get()
             );

